### ISSUE TITLE
DLT_v2  question 

### ISSUE THREAD

**Issue by tomjerrygithub** [stmoon/DeepImageStitching#4]
Hi ~ 
# Compute h33
    try:
        _ ,_, V = torch.svd(A.detach(),some=False) # Not compute gradient of svd
    except:
        print(A)
        raise RuntimeError
    h33 = V[:,8,8]

raise RuntimeError  how to deal with this ?

**Comment by michael-1003** [comment]
Thanks for asking. Actually, I skipped that part since this project has time limitation.
To answer your question, that exception occurs when matrix A is singular when computing svd of A.
I may suggest how to deal with it. 
A matrix become singular when 4 sampled point is formed as triangle like figure below.
1
│  ＼
│ 　 3
│　　 ＼   
2─────4
To solve this case, maybe check 3rd point's position and if it is on line 1-4, reinforce 3rd point should have such 1or 2 pixel offset from its original sampled point.
I remember this cases frequently happens when setting learning rate higher than 1e-2(or 3?).
To avoid this, small learning rate should be used.
However, using small learning rate cannot estimate when two image has great offset, so we are on discussion to solve this problem.
We consider to update this part later since this repository should be freeze until end of this month.

**Comment by michael-1003** [comment]
Also, one more thing, if pytorch version is low, it does not return proper result of batch-wise svd.
So please use latest version of pytorch.

**Comment by tomjerrygithub** [comment]
i tried a batch svd  but failed too    
https://github.com/KinglittleQ/torch-batch-svd   
run but  loss become nan  immediately

**Comment by michael-1003** [comment]
No, what I mean was to use **torch.svd function in updated pytorch version(we tested on 1.3.1).**
In our implementation, torch.svd option should be set as **some=False**, to use 9th column in computed V matrix (size B*9*9, B is batch size).
However, the function you linked is equivalent as torch.svd(x, **some=True,** compute_uv=True), which ignore last singular value and returns V matrix having size of B*8*8, not contains 9th column.
In that case I think h33 = V[:,8,8] part will return out of index error and running stops.

To make problem clear, what kind of failure did you get?
1) Out of index error
2) Runtime Error by Matrix A is singular
3) Code runs but nan loss occured

**Comment by tomjerrygithub** [comment]
Thank you for your detailed reply.
My pytorch version is 1.1.0 

And i use  torch.svd   raise Error  directly
Then i use the https://github.com/KinglittleQ/torch-batch-svd  code runs but nan loss occured。

So， you mean use torch.svd  some=False  can run batch  in  1.3.1  pytorch version ?

i want to train a model to  estimate homography matirx for AR  using  unsupervised method

**Comment by tomjerrygithub** [comment]
torch 1.3.1 ok   torch.svd  worked  
i use the  DLT and  STN part  to do  unsupervised learning

**Comment by michael-1003** [comment]
Good to hear that it worked. Thanks.

**Comment by tomjerrygithub** [comment]
Do you read the paper  Content-Aware Unsupervised Deep Homography Estimation
https://arxiv.org/abs/1909.05983 
i think it's a better way to estimate homography   but not opensource now

**Comment by michael-1003** [comment]
Yes. we read that paper and later we will implement it. Thanks for comments.
